model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java

```diff
@@ -0,0 +1,140 @@
+                        cb.equal(root.get(modelField.getName()), value[0])
+                    );
+                } else if (modelField.equals(SearchableFields.IS_CLIENT_ROLE)) {
+                    validateValue(value, modelField, op, Boolean.class);
+
+                    return new JpaRoleModelCriteriaBuilder((cb, root) -> 
+                        cb.equal(
+                            cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fClientRole")), 
+                            cb.literal(convertToJson(value[0]))
+                        )
+                    );
+                } else {
+                    throw new CriterionNotSupportedException(modelField, op);
+                }
+            case IN:
+                if (modelField.equals(SearchableFields.ID)) {
+                    validateValue(value, modelField, op, String.class);
+
+                    return new JpaRoleModelCriteriaBuilder((cb, root) -> {
+                        In<String> in = cb.in(root.get("id"));
+                        ((Stream<String>) value[0]).forEach(id -> in.value(id));
+                        return in;
+                    });
+                } else {
+                    throw new CriterionNotSupportedException(modelField, op);
+                }
+            case ILIKE:
+                if (modelField.equals(SearchableFields.NAME) ||
+                    modelField.equals(SearchableFields.DESCRIPTION)) {
+
+                    validateValue(value, modelField, op, String.class);
+
+                    return new JpaRoleModelCriteriaBuilder((cb, root) ->
+                        cb.like(
+                            cb.lower(cb.function("->", String.class, root.get("metadata"), cb.literal(getFieldName(modelField)))),
+                            cb.literal(convertToJson(value[0]).toLowerCase())
+                        )
+                    );
+                } else {
+                    throw new CriterionNotSupportedException(modelField, op);
+                }
+            default:
+                throw new CriterionNotSupportedException(modelField, op);
+        }
+    }
+
+    @Override
+    public JpaRoleModelCriteriaBuilder and(JpaRoleModelCriteriaBuilder... builders) {
+        return new JpaRoleModelCriteriaBuilder(
+            (cb, root) -> cb.and(Stream.of(builders).map(b -> b.getPredicateFunc().apply(cb, root)).toArray(Predicate[]::new))
+        );
+    }
+
+    @Override
+    public JpaRoleModelCriteriaBuilder or(JpaRoleModelCriteriaBuilder... builders) {
+        return new JpaRoleModelCriteriaBuilder(
+            (cb, root) -> cb.or(Stream.of(builders).map(b -> b.getPredicateFunc().apply(cb, root)).toArray(Predicate[]::new))
+        );
+    }
+
+    @Override
+    public JpaRoleModelCriteriaBuilder not(JpaRoleModelCriteriaBuilder builder) {
+        return new JpaRoleModelCriteriaBuilder(
+            (cb, root) -> cb.not(builder.getPredicateFunc().apply(cb, root))
+        );
+    }
```

Comment:
Could these methods be defined in `JpaModelCriteriaBuilder`?
Perhaps if the class was updated like this:
```java
public class JpaModelCriteriaBuilder<Self> {
    private static final Function<BiFunction<CriteriaBuilder, Root<JpaRoleEntity>, Predicate>, Self> instantiator;

   // ... e.g.
    @Override
    public Self not(Self builder) {
        return instantiator.apply(
            (cb, root) -> cb.not(builder.getPredicateFunc().apply(cb, root))
        );
    }
```

and this `instantiator` could be initialized in the constructor, called from the descendants like this:

```
    JpaRoleModelCriteriaBuilder() {
        super(JpaRoleModelCriteriaBuilder::new);
    }
```

Reply:
Thank you for the suggestion. I've played a little bit with it and came up with 2 possible solutions (https://github.com/keycloak/keycloak/pull/9415/commits/28dacf606aa1d5ad699d844f050c7d0979e706aa or https://github.com/keycloak/keycloak/pull/9415/commits/80d13940a7cfbabc295b1762800ab08169e0b8dd). 

The first approach uses `public abstract class JpaModelCriteriaBuilder<E, Self extends JpaModelCriteriaBuilder>` and child classes then `implements ModelCriteriaBuilder`. 
Methods `and`, `or` and `not` don't require casting `Self builder` into `JpaModelCriteriaBuilder` using this approach. But we cannot mark the methods by `@Override` annotation (and IDE doesn't find then as implementations, at least in my case :) )

While he second approach uses `public abstract class JpaModelCriteriaBuilder<E, M, Self extends ModelCriteriaBuilder<M, Self>> implements ModelCriteriaBuilder<M, Self>` and child classes then just extends it. 

I wonder which approach would be better, wdyt? There might be some other even better solution though.

Reply:
https://github.com/keycloak/keycloak/commit/80d13940a7cfbabc295b1762800ab08169e0b8dd is the correct one.

Reply:
Thank you